### ISSUE TITLE
Zero-Copy replication (experimental feature not for production usage) on remote disk lost data during mutation

### ISSUE THREAD

**Issue by ianton-ru** [ClickHouse/ClickHouse#39560]
Scenario for reproduce:

ReplicatedMergeTree table with parts on remote disk with enabled zero-copy replication

1. replica1 creates new part on remote disk, creates objects on remote disk and local metadata files with references on objects
2. replica2 fetches this part via zero-copy, creates local metadata files with references on same objects
3. replica1 executes mutation, during mutation creates new part with same data. For this replica1 creates hardlinks on metadata files, in metadata increments ref_count counter (counter of local links on that metadata file)
4. replica2 fetches new part via zero-copy. It's has no information that it already has one copy of metadata, so creates new metadata files with ref_count=0, old metadata files still have ref_count=0 too.
5. replica1 removes old part. It sees than ref_count>0, decrements it and leaves objects on remote disk.
6. replica2 removes old part. It gets information form zero-copy mechanism that this is last copy of old part on cluster, sees that ref_count=0 in local metadata files and removes remote objects.

After that replicas got errors like in #[37791](https://github.com/ClickHouse/ClickHouse/issues/37791)

Possible ways to fix it:

1. Bad and slow - make a fullscan for all local metadata files to find duplicates and create hardlinks instead. This process should start too often - after every part creation or before removing.
2. With large memory usage - create a revert index in memory "remote object"->"local metadata files" to fast find existing local copies for metadata files
3. With local disk usage - for every metadata file create one additional hardlink in special folder inside /var/lib/clickhouse/disks/<remote_disk>/<new_special_folder_for_metadata_hardlinks>, where link names associated with remote object names (hash of remote object name for example).

@alesapin @kssenii Please note, because this looks like a serious bug in zero-copy logic. :(

I think that variant 3 is better, but it's not perfect.

**Comment by alexey-milovidov** [comment]
It looks similar to https://github.com/ClickHouse/ClickHouse/pull/36299

**Comment by alesapin** [comment]
It's exactly the case @alexey-milovidov and it's already fixed. @ianton-ru check the PR.

**Comment by alexey-milovidov** [comment]
@orloffv "Zero-copy replication" is not a production-ready feature.
It was not finished by contributors, and I don't recommend using it. Even in the latest release, there are known issues.
For example, the developers did not implement support for projections yet.

If you are using "zero-copy replication" for real users, it is dangerous.

**Comment by ianton-ru** [comment]
@alesapin Still not fixed :(

Steps to reproduce

```create table test on cluster '{cluster}' (key UInt32, value1 String, value2 String, value3 String) engine=ReplicatedMergeTree('/test', '{replica}') order by key partition by (key % 4) settings storage_policy='object_storage'

insert into test select * from generateRandom('key UInt32, value1 String, value2 String, value3 String') limit 1000000

alter table test add column valueX String materialized value1

alter table test materialize column valueX
```

after that fast (in 8 minutes, before clickhouse removes original parts) checks on both replicas

```grep -A 1 r00000000000000000000 -R /var/lib/clickhouse/disks/object_storage/store/ | grep -B 1 "\-0" | grep r00000000000000000000 | sort -k 2 | uniq -df 1
```

This found non-uniq referrrers with ref_count==0

```var/lib/clickhouse/disks/object_storage/store/bbb/bbb323bd-fa3d-421b-b855-9be3c96b4344/0_0_0_0_1/primary.idx:128	r0000000000000000000000000000000000000000000000000000000000000010-file-sqcxsjospyhkqpppcltamahozreikqgt
/var/lib/clickhouse/disks/object_storage/store/bbb/bbb323bd-fa3d-421b-b855-9be3c96b4344/0_0_0_0_1/key.bin:1005515	r0000000000000000000000000000000000000000000000000000000000000011-file-vxvqlygfprntkjpjabgewywzfsxkpejd
/var/lib/clickhouse/disks/object_storage/store/bbb/bbb323bd-fa3d-421b-b855-9be3c96b4344/0_0_0_0_1/key.mrk2:768	r0000000000000000000000000000000000000000000000000000000000000100-file-foccgmbwltdjzgquabqkwbdjmchebdmt
...
```

```
# cat /var/lib/clickhouse/disks/object_storage/store/bbb/bbb323bd-fa3d-421b-b855-9be3c96b4344/0_0_0_0_1/primary.idx
3
1	128
128	r0000000000000000000000000000000000000000000000000000000000000010-file-sqcxsjospyhkqpppcltamahozreikqgt
0
0

# cat /var/lib/clickhouse/disks/object_storage/store/bbb/bbb323bd-fa3d-421b-b855-9be3c96b4344/0_0_0_0/primary.idx
3
1	128
128	r0000000000000000000000000000000000000000000000000000000000000010-file-sqcxsjospyhkqpppcltamahozreikqgt
0
0
```

I checked on version 22.7.1.2484 and from master branch (commit 9525bd19bf785953db260df8ef4a17231d90a881)

**Comment by zhaojintaozhao** [comment]
Hello, @ianton-ru, @alexey-milovidov 

After I test your example test case, I can also find data in ClickHouse. I think this problem has been solved by #36299.
In my test system, there are 2 replicas. All test steps are all the same as you. 1st, create new ReplicatedMergeTree table on tow cluster. 2nd, insert data to this table. 3rd 
alter this table and add new column. 4th alter this table and materialize new column. 5th after 20 min, data also can be queried from ClickHouse.

I have analyzed the code and the log of my test system.
1st I find the refcount of the "value1.bin" flle in first replica is 1. But the refcount of the "value1.bin" flle in first replica is  0.
2nd. Clickhouse set the hardlink file name in old part of zero copy path in ZooKeeper. I get the hardlink files(''value1.bin", "value1.mr2" and so on) in 
zk zero copy lock path: /clickhouse/zero_copy/zero_copy_s3/{table_id}/{part_name}. We can analyze the logic from createZeroCopyLockNode func of StorageReplicatedMergeTree.cpp.
3. When ClickHouse delete old part, ClickHouse will get hardlink filenames from zk zero copy lock path. ClickHouse will skip these hardlink files when remove shared files. We can analyze the logic from unlockSharedDataByID func of StorageReplicatedMergeTree.cpp. The unlockSharedDataByID func return not removed files.

**Comment by zhaojintaozhao** [comment]
Hi  @ianton-ru , @orloffv :  Please confirm your test case again. You can check the remote file on the remote fs and get the hardlink files of zero-copy path on the zookeeper.

**Comment by ianton-ru** [comment]
@zhaojintaozhao I check it again, you are right. Before I check only that ref_count in metadata has incorrect value, but fix #36299 uses other mechanics - it make a record in Keeper for old part that that some of objects have hardlinks, and do not removed that objects during part cleanup. So in my sample objects keeped on remote storage.
I need to think more, for me still looks dangerous, depends on part deletion order.
